### PR TITLE
stats: implement custom metrics

### PR DESCRIPTION
**Pull request by kyessenov**
Signed-off-by: Kuat Yessenov <kuat@google.com>

**What this PR does / why we need it**:

**Which issue this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*: fixes #

**Special notes for your reviewer**:

### CODE REVIEW

**Review by lei-tang — COMMENTED** [review]
Do you have a document describing the workflow of this feature and how to use this feature?

**Comment by kyessenov** [comment]
@lei-tang This is the last gap with Wasm stats, see the reference https://istio.io/latest/docs/reference/config/proxy_extensions/stats/. It's not exposed in Telemetry API, but we'll switch EnvoyFilter users to native stats, so need this.

**Review by lei-tang — APPROVED** [review]
Thanks for your PR, Kuat!

File: extensions/stats/config.proto [hunk 1]
```diff
@@ -64,7 +64,7 @@ message MetricDefinition {
   // Metric value expression.
   string value = 2;
 
-  // NOT IMPLEMENTED (Optional) Metric type.
```
Comment (zirain) [inline extensions/stats/config.proto:67]: need update doc

File: source/extensions/filters/http/istio_stats/istio_stats.cc [hunk 2]
```diff
@@ -481,6 +511,37 @@ struct Config : public Logger::Loggable<Logger::Id::filter> {
...
+          case stats::MetricType::GAUGE:
+            metric_type = MetricOverrides::MetricType::Gauge;
+            break;
+          case stats::MetricType::HISTOGRAM:
+            metric_type = MetricOverrides::MetricType::Histogram;
+            break;
+          default:
+            break;
+        }
+        metric_overrides_->custom_metrics_.try_emplace(
```
Comment (zirain) [inline source/extensions/filters/http/istio_stats/istio_stats.cc:539]: I try to do this myself but get stuck here, I learn a lot from this PR, thanks!

**Pull request merged**